### PR TITLE
sql: Update reg* casting to support invalid DBs

### LINKED ISSUE #25080: `has_table_privilege` on internal system tables fails when schema/database is not set
### What version of Materialize are you using?

v0.86.1

### What you expect to happen

`select has_table_privilege('mz_internal.mz_activity_log', 'select');` returns a result:
Screenshot 2024-02-08 at 11 47 47 AM

### What actually happens
Screenshot 2024-02-08 at 11 48 17 AM

### Steps to reproduce:
1. Set your schema and database to falsy values like `null` 
2. Run `select has_table_privilege('mz_internal.mz_activity_log', 'select');`

### PR DESCRIPTION
**Pull request by jkosh44**
This commit fixes text to reg* casting (such as text to regclass) so that items in a system schema can be correctly resolved when the active database is invalid.

Fixes #25080

### Motivation
This PR fixes a recognized bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Allow users to cast text to `reg*` types (such as `regclass`) for items in system schemas when the active database is invalid. This also fixes any function that uses `reg*` casting such as access privilege inquiry functions.

### CODE REVIEW

File: src/sql/src/func.rs [hunk 1]
```diff
@@ -3927,7 +3927,14 @@ pub static MZ_INTERNAL_BUILTINS: Lazy<BTreeMap<&'static str, Func>> = Lazy::new(
...
                     ON search_schema.name = s.name
                 JOIN
-                    mz_catalog.mz_databases AS d
+                    (
+                        SELECT id, name FROM mz_catalog.mz_databases
+                        -- If the provided database does not exist, add a row for it so that it
+                        -- can still join against ambient schemas.
+                        UNION ALL
+                        SELECT '', $1
+                        WHERE (SELECT $1 NOT IN (SELECT name FROM mz_catalog.mz_databases))
```
Comment (jkosh44) [inline src/sql/src/func.rs:3936]: This can probably actually be simplified to `WHERE $1 NOT IN (SELECT name FROM mz_catalog.mz_databases)`

Comment (jkosh44) [reply src/sql/src/func.rs]: Done.

**Review by sploiselle — COMMENTED** [review]
Can we get a test for this behavior on user objects, i.e. something that doesn't use an ambient database?

**Comment by jkosh44** [comment]
> Can we get a test for this behavior on user objects, i.e. something that doesn't use an ambient database?

Done!

**Pull request merged**